includes/gateway/class-omise-payment-creditcard.php

```diff
@@ -166,243 +166,192 @@ public function payment_fields() {
+			if ( Omise_Charge::is_failed( $charge ) ) {
+				return $this->payment_failed( Omise_Charge::get_error_message( $charge ) );
+			}
 
-						default:
-							// Default behaviour is, check if it paid first.
-							$success = Omise_Charge::is_paid( $charge );
+			if ( $this->omise_3ds ) {
+				$order->add_order_note(
+					sprintf(
+						__( 'Omise: Processing with a 3-D Secure payment, redirecting buyer out to %s', 'omise' ),
+						esc_url( $charge['authorize_uri'] )
+					)
+				);
 
-							// Then, check is authorized after if the first condition is false.
-							if ( ! $success )
-								$success = Omise_Charge::is_authorized( $charge );
+				return array(
+					'result'   => 'success',
+					'redirect' => $charge['authorize_uri'],
+				);
+			}
 
-							break;
+			switch ( strtoupper( $this->payment_action ) ) {
+				case 'MANUAL_CAPTURE':
+					$success = Omise_Charge::is_authorized( $charge );
+					if ( $success ) {
+						$order->add_order_note(
+							sprintf(
+								wp_kses(
+									__( 'Omise: Payment processing.<br/>An amount %1$s %2$s has been authorized', 'omise' ),
+									array( 'br' => array() )
+								),
+								$order->get_total(),
+								$order->get_order_currency()
+							)
+						);
 					}
 
-					if ( ! $success ) {
-						throw new Exception( __( 'Note that your payment might already has been processed. Please contact our support team if you have any questions.', 'omise' ) );
+					break;
+
+				case 'AUTO_CAPTURE':
+					$success = Omise_Charge::is_paid( $charge );
+					if ( $success ) {
+						$order->payment_complete();
+						$order->add_order_note(
+							sprintf(
+								wp_kses(
+									__( 'Omise: Payment successful.<br/>An amount %1$s %2$s has been paid', 'omise' ),
+									array( 'br' => array() )
+								),
+								$order->get_total(),
+								$order->get_order_currency()
+							)
+						);
 					}
 
-					// Remove cart
-					WC()->cart->empty_cart();
-					return array (
-						'result'   => 'success',
-						'redirect' => $this->get_return_url( $order )
-					);
-				}
-			} catch( Exception $e ) {
-				wc_add_notice(
-					sprintf(
-						wp_kses(
-							__( 'Seems we cannot process your payment properly:<br/>%s', 'omise' ),
-							array( 'br' => array() )
-						),
-						$e->getMessage()
-					),
-					'error'
-				);
+					break;
 
-				$order->add_order_note(
-					sprintf(
-						__( 'Omise: Payment failed, %s', 'omise' ),
-						$e->getMessage()
-					)
-				);
+				default:
+					// Default behaviour is, check if it paid first.
+					$success = Omise_Charge::is_paid( $charge );
 
-				return;
+					// Then, check is authorized after if the first condition is false.
+					if ( ! $success )
+						$success = Omise_Charge::is_authorized( $charge );
+
+					break;
+			}
+
+			if ( ! $success ) {
+				return $this->payment_failed( __( 'Note that your payment might already has been processed. Please contact our support team if you have any questions.', 'omise' ) );
```

Comment:
Note that your payment may have already been processed. Please contact our support team if you have any questions.